### ISSUE TITLE
Unsigned Dokan. Fails if any other version of Dokan is installed.

### ISSUE THREAD

**Issue by vagaerg** [keybase/client#3985]
I am trying to install keybase client (latest from the downloads website, 1.0.16.79) on my Windows 10 Pro machine (x64). If Dokan is not installed, Windows will complain about an unsigned driver being installed in the system. If Dokan is installed (a version from their git page, correctly signed and that installs correctly), Keybase installer will fail with this log:

`
[0794:078C][2016-08-21T14:14:07]i001: Burn v3.10.2.2516, Windows v10.0 (Build 14393: Service Pack 0), path: C:\Users\Pablo\AppData\Local\Temp\{C5C6AA59-49E9-434C-B4AC-2E0DEED4E3FD}\.cr\keybase_setup_386 (2).exe
[0794:078C][2016-08-21T14:14:07]i000: Initializing string variable 'DokanProduct64' to value '{65A3A964-3DC3-0100-0000-160621082245}'
[0794:078C][2016-08-21T14:14:07]i000: Initializing string variable 'DokanProduct86' to value '{65A3A986-3DC3-0100-0000-160621082245}'
[0794:078C][2016-08-21T14:14:07]i009: Command Line: '"-burn.clean.room=C:\Users\Pablo\Downloads\keybase_setup_386 (2).exe"'
[0794:078C][2016-08-21T14:14:07]i000: Setting string variable 'WixBundleOriginalSource' to value 'C:\Users\Pablo\Downloads\keybase_setup_386 (2).exe'
[0794:078C][2016-08-21T14:14:07]i000: Setting string variable 'WixBundleOriginalSourceFolder' to value 'C:\Users\Pablo\Downloads\'
[0794:078C][2016-08-21T14:14:07]i000: Setting string variable 'LOGPATH_PROP' to value 'C:\Users\Pablo\AppData\Local\Temp\Keybase_20160821141407.log'
[0794:078C][2016-08-21T14:14:07]i000: Setting string variable 'WixBundleName' to value 'Keybase'
[0794:078C][2016-08-21T14:14:07]i000: Setting string variable 'WixBundleManufacturer' to value 'Keybase, Inc.'
[0794:0860][2016-08-21T14:14:07]i000: Setting numeric variable 'WixStdBALanguageId' to value 1033
[0794:0860][2016-08-21T14:14:07]i000: Setting version variable 'WixBundleFileVersion' to value '1.0.16.79'
[0794:078C][2016-08-21T14:14:07]i100: Detect begin, 3 packages
[0794:078C][2016-08-21T14:14:07]i000: Registry key not found. Key = 'SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{65A3A986-3DC3-0100-0000-160621082245}'
[0794:078C][2016-08-21T14:14:07]i052: Condition 'NOT DokanUninstallString' evaluates to true.
[0794:078C][2016-08-21T14:14:07]i000: Registry key not found. Key = 'SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{65A3A986-3DC3-0100-0000-160621082245}'
[0794:078C][2016-08-21T14:14:07]i052: Condition 'NOT DokanUninstallString' evaluates to true.
[0794:078C][2016-08-21T14:14:07]i000: Registry key not found. Key = 'SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{65A3A964-3DC3-0100-0000-160621082245}'
[0794:078C][2016-08-21T14:14:07]i052: Condition 'NOT DokanUninstallString' evaluates to true.
[0794:078C][2016-08-21T14:14:07]i000: Registry key not found. Key = 'SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{65A3A964-3DC3-0100-0000-160621082245}'
[0794:078C][2016-08-21T14:14:07]i000: Registry key not found. Key = 'Software\Microsoft\Windows\CurrentVersion\Uninstall\{70E747DE-4E09-44B0-ACAD-784AA9D79C02}_is1'
[0794:078C][2016-08-21T14:14:07]i000: Setting numeric variable 'InnoCLIInstalled' to value 0
[0794:078C][2016-08-21T14:14:07]i000: Registry key not found. Key = 'Software\Microsoft\Windows\CurrentVersion\Uninstall\{70E747DE-4E09-44B0-ACAD-784AA9D79C02}_is1'
[0794:078C][2016-08-21T14:14:07]i000: Setting numeric variable 'InnoCLIInstalled64' to value 0
[0794:078C][2016-08-21T14:14:07]i000: Registry key not found. Key = 'SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{357F272E-BE0E-409F-8E39-0BB9827F5716}_is1'
[0794:078C][2016-08-21T14:14:07]i052: Condition 'NOT InnoUninstallString' evaluates to true.
[0794:078C][2016-08-21T14:14:07]i000: Registry key not found. Key = 'SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{357F272E-BE0E-409F-8E39-0BB9827F5716}_is1'
[0794:078C][2016-08-21T14:14:07]i052: Condition 'NOT InnoUninstallString' evaluates to true.
[0794:078C][2016-08-21T14:14:07]i000: Registry key not found. Key = 'SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{DEB2E54C-C39F-4DC8-93A7-ABE0AB91DDCA}_is1'
[0794:078C][2016-08-21T14:14:07]i052: Condition 'NOT InnoUninstallString' evaluates to true.
[0794:078C][2016-08-21T14:14:07]i000: Registry key not found. Key = 'SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{DEB2E54C-C39F-4DC8-93A7-ABE0AB91DDCA}_is1'
[0794:078C][2016-08-21T14:14:07]i052: Condition 'NOT InnoUninstallString' evaluates to true.
[0794:078C][2016-08-21T14:14:07]i000: Registry key not found. Key = 'SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{1B2672D9-2BAD-4C11-BA53-A75AF6FD7789}_is1'
[0794:078C][2016-08-21T14:14:07]i052: Condition 'NOT InnoUninstallString' evaluates to true.
[0794:078C][2016-08-21T14:14:07]i000: Registry key not found. Key = 'SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{1B2672D9-2BAD-4C11-BA53-A75AF6FD7789}_is1'
[0794:078C][2016-08-21T14:14:07]i000: Registry key not found. Key = 'SOFTWARE\Keybase'
[0794:078C][2016-08-21T14:14:07]i052: Condition 'NOT TargetDokanUninstallKey' evaluates to true.
[0794:078C][2016-08-21T14:14:07]i000: Registry key not found. Key = 'SOFTWARE\Keybase'
[0794:078C][2016-08-21T14:14:07]i000: Setting version variable 'WINTRUSTVERSION' to value '10.0.14393.0'
[0794:078C][2016-08-21T14:14:07]i052: Condition 'NOT InnoUninstallString' evaluates to true.
[0794:078C][2016-08-21T14:14:07]i052: Condition 'DokanUninstallString and NOT (WixBundleAction=3 AND ( TargetDokanUninstallKey <> DokanProduct64 AND TargetDokanUninstallKey <> DokanProduct86 ))' evaluates to false.
[0794:078C][2016-08-21T14:14:07]i101: Detected package: runquiet.exe, state: Present, cached: None
[0794:078C][2016-08-21T14:14:07]i101: Detected package: DokanSetup_redist.exe, state: Absent, cached: Complete
[0794:078C][2016-08-21T14:14:07]i101: Detected package: KeybasePrograms, state: Absent, cached: Complete
[0794:078C][2016-08-21T14:14:07]i052: Condition 'Installed OR VersionNT > v6.1 OR (VersionNT = v6.1 AND WINTRUSTVERSION >= v6.1.7601.18741)' evaluates to true.
[0794:078C][2016-08-21T14:14:07]i052: Condition 'NOT InnoCLIInstalled AND NOT InnoCLIInstalled64' evaluates to true.
[0794:078C][2016-08-21T14:14:07]i199: Detect complete, result: 0x0
[0794:078C][2016-08-21T14:14:12]i200: Plan begin, 3 packages, action: Install
[0794:078C][2016-08-21T14:14:12]w321: Skipping dependency registration on package with no dependency providers: runquiet.exe
[0794:078C][2016-08-21T14:14:12]w321: Skipping dependency registration on package with no dependency providers: DokanSetup_redist.exe
[0794:078C][2016-08-21T14:14:12]i000: Setting string variable 'WixBundleLog_DokanSetup_redist.exe' to value 'C:\Users\Pablo\AppData\Local\Temp\Keybase_20160821141407_000_DokanSetup_redist.exe.log'
[0794:078C][2016-08-21T14:14:12]i000: Setting string variable 'WixBundleRollbackLog_KeybasePrograms' to value 'C:\Users\Pablo\AppData\Local\Temp\Keybase_20160821141407_001_KeybasePrograms_rollback.log'
[0794:078C][2016-08-21T14:14:12]i000: Setting string variable 'WixBundleLog_KeybasePrograms' to value 'C:\Users\Pablo\AppData\Local\Temp\Keybase_20160821141407_001_KeybasePrograms.log'
[0794:078C][2016-08-21T14:14:12]i201: Planned package: runquiet.exe, state: Present, default requested: Present, ba requested: Present, execute: None, rollback: None, cache: No, uncache: No, dependency: None
[0794:078C][2016-08-21T14:14:12]i201: Planned package: DokanSetup_redist.exe, state: Absent, default requested: Present, ba requested: Present, execute: Install, rollback: None, cache: No, uncache: No, dependency: None
[0794:078C][2016-08-21T14:14:12]i201: Planned package: KeybasePrograms, state: Absent, default requested: Present, ba requested: Present, execute: Install, rollback: Uninstall, cache: No, uncache: No, dependency: Register
[0794:078C][2016-08-21T14:14:12]i299: Plan complete, result: 0x0
[0794:078C][2016-08-21T14:14:12]i300: Apply begin
[0794:078C][2016-08-21T14:14:12]i010: Launching elevated engine process.
[0794:078C][2016-08-21T14:14:13]i011: Launched elevated engine process.
[0794:078C][2016-08-21T14:14:13]i012: Connected to elevated engine.
[13D4:0E0C][2016-08-21T14:14:13]i358: Pausing automatic updates.
[13D4:0E0C][2016-08-21T14:14:13]i359: Paused automatic updates.
[13D4:0E0C][2016-08-21T14:14:13]i360: Creating a system restore point.
[13D4:0E0C][2016-08-21T14:14:13]i361: Created a system restore point.
[0794:078C][2016-08-21T14:14:13]i370: Session begin, registration key: SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{445406b3-3d58-4ff0-8c26-ad704a1b95ac}, options: 0x7, disable resume: No
[0794:078C][2016-08-21T14:14:13]i000: Caching bundle from: 'C:\Users\Pablo\AppData\Local\Temp\{C340928A-AEE1-45C9-A977-1C7B4E5DB1AB}\.be\Keybase_1.0.16-20160713110422+a1c00b3.386.exe' to: 'C:\Users\Pablo\AppData\Local\Package Cache\{445406b3-3d58-4ff0-8c26-ad704a1b95ac}\Keybase_1.0.16-20160713110422+a1c00b3.386.exe'
[0794:078C][2016-08-21T14:14:13]i320: Registering bundle dependency provider: {445406b3-3d58-4ff0-8c26-ad704a1b95ac}, version: 1.0.16.79
[0794:078C][2016-08-21T14:14:13]i371: Updating session, registration key: SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{445406b3-3d58-4ff0-8c26-ad704a1b95ac}, resume: Active, restart initiated: No, disable resume: No
[13D4:1718][2016-08-21T14:14:14]i304: Verified existing payload: DokanSetup_redist.exe at path: C:\ProgramData\Package Cache\2FC376A125AB31E0BA39FC5C83ED7905A36BF2B9\DokanSetup_redist.exe.
[0794:19C4][2016-08-21T14:14:14]i304: Verified existing payload: KeybasePrograms at path: C:\Users\Pablo\AppData\Local\Package Cache\{803E48B3-ECDB-4B7D-9795-63FAD60A2E26}v1.0.16.79\Keybase.msi.
[13D4:0E0C][2016-08-21T14:14:14]i301: Applying execute package: DokanSetup_redist.exe, action: Install, path: C:\ProgramData\Package Cache\2FC376A125AB31E0BA39FC5C83ED7905A36BF2B9\DokanSetup_redist.exe, arguments: '"C:\ProgramData\Package Cache\2FC376A125AB31E0BA39FC5C83ED7905A36BF2B9\DokanSetup_redist.exe" /quiet'
[13D4:0E0C][2016-08-21T14:14:15]e000: Error 0x80070666: Process returned error: 0x666
[13D4:0E0C][2016-08-21T14:14:15]e000: Error 0x80070666: Failed to execute EXE package.
[0794:078C][2016-08-21T14:14:15]e000: Error 0x80070666: Failed to configure per-machine EXE package.
[0794:078C][2016-08-21T14:14:15]i319: Applied execute package: DokanSetup_redist.exe, result: 0x80070666, restart: None
[0794:078C][2016-08-21T14:14:15]e000: Error 0x80070666: Failed to execute EXE package.
[0794:078C][2016-08-21T14:14:15]i372: Session end, registration key: SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{445406b3-3d58-4ff0-8c26-ad704a1b95ac}, resume: None, restart: None, disable resume: No
[0794:078C][2016-08-21T14:14:15]i330: Removed bundle dependency provider: {445406b3-3d58-4ff0-8c26-ad704a1b95ac}
[0794:078C][2016-08-21T14:14:15]i352: Removing cached bundle: {445406b3-3d58-4ff0-8c26-ad704a1b95ac}, from path: C:\Users\Pablo\AppData\Local\Package Cache\{445406b3-3d58-4ff0-8c26-ad704a1b95ac}\
[0794:078C][2016-08-21T14:14:15]i371: Updating session, registration key: SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{445406b3-3d58-4ff0-8c26-ad704a1b95ac}, resume: None, restart initiated: No, disable resume: No
[0794:078C][2016-08-21T14:14:15]i399: Apply complete, result: 0x80070666, restart: None, ba requested restart:  No
`

I have also tried 1.0.17-20160820134520+9aefaa4.386. With this version, it installs correctly but kbfs does not work.

**Comment by zanderz** [comment]
Microsoft has announced that fresh installs of Win 10 Anniversary Update will enforce a requirement that kernel drivers must first be signed by Microsoft. We are still in the process of submitting drivers, but what is puzzling is that I know Dokany hasn't completed that process either, so I have no idea how that would install correctly. Can you attach all the log files in your `%temp%` directory that have dokan in the name? Maybe that package had trouble installing a prerequisite?

**Comment by vagaerg** [comment]
Here are all the logs related to Dokan in my %temp% folder. It is very strange then! I could install it and windows did not say anything, unlike with the Keybase.io client.

Thank you very much

**Comment by zanderz** [comment]
Thanks. That shows a bunch of installing and uninstalling each Dokan package; the one we shipped on July 13 and the latest dokan-dev one. The only failure was when the newere dokan was already present on your machine, otherwise they both seem to think installation succeeded and that rebooting would be required next - were you prompted to reboot? Neither of these seem to show that they were blocked by Windows. What error did you get exactly, and how do you know the dokan-dev one worked?

FYI, it should be ok to remove our Dokan and install theirs. Your KBFS mount should then come back after reboot.

```
MSI (s) (1C:D0) [13:48:48:384]: Product: Dokan Library 1.0.0.26 (x64) -- Installation completed successfully.

MSI (s) (1C:D0) [13:48:48:399]: Windows Installer installed the product. Product Name: Dokan Library 1.0.0.26 (x64). Product Version: 1.0.0.26. Product Language: 1033. Manufacturer: Dokany Project. Installation success or error status: 0.

MSI (s) (1C:D0) [13:48:48:399]: Value of RebootAction property is 1
MSI (s) (1C:D0) [13:48:48:399]: Windows Installer requires a system restart. Product Name: Dokan Library 1.0.0.26 (x64). Product Version: 1.0.0.26. Product Language: 1033. Manufacturer: Dokany Project. Type of System Restart: 2. Reason for Restart: 2.

MSI (s) (1C:D0) [13:48:48:399]: Product: Dokan Library 1.0.0.26 (x64). Restart required. The product installation or update required a restart for all changes to take effect.  The restart was deferred to a later time.
```

**Comment by zanderz** [comment]
I just tested a recent Keybase build with a fresh install of that same build of windows 10 (14393), and didn't get any driver blockage. If you like, you can try https://prerelease.keybase.io/windows/Keybase_1.0.17-20160822152935%2B692af12.386.exe

**Comment by vagaerg** [comment]
Thank you very much for your reply. It rebooted promptly (see issue #3930 ) when I downloaded the version from your page, but I could stop it so the installation finished properly. Dokan showed up in the "Remove programs" part of the control panel, so it was installed, but I got a message saying (something similar to) "Windows needs drivers to be signed", and KBFS did not work.

If I download the prerelease version I posted on my first message, it didn't reboot promptly but it still showed the message about the drivers not being signed. 

If I installed Dokan after keybase, no errors or messages but KBFS would not work.

I will try the version you linked as soon as I get back home

Thank you

**Comment by vagaerg** [comment]
The version you proposed is having the same issue. Maybe you tried it on a 32 bit version of Windows? (as they don't block unsigned drivers)
![2016-08-24 10_08_22-pinentry](https://cloud.githubusercontent.com/assets/1702993/17924982/e810783c-69e2-11e6-85fe-646d92fa790a.png)

**Comment by gbaker-blackbirdit** [comment]
@zanderz Does your Windows 10 machine have UEFI with secure boot enabled?  If not, that might explain the difference in your experience.  I believe Windows 10 still allows boot through BIOS.

**Comment by AArnott** [comment]
Why does the Keybase client even install a system driver? That is unclear and not explained during setup. It actually leads me to distrust the client software (like it wants to root-kit me or something) when it wants to install a system driver. Heck, the install shouldn't even require elevation if it's just perform signing for me with my own PGP key.

**Comment by maxtaco** [comment]
@AArnott our next release will make install of KBFS optional. We need a kernel module on MacOS and Windows for this feature (it comes standard on Linux).  So in that release, the install will be unprivileged unless you choose to enable KBFS, which will escalate privileges.

**Comment by AArnott** [comment]
Thanks. That sounds great. What is KBFS, BTW?

**Comment by vagaerg** [comment]
@AArnott Keybase Filesystem
@maxtaco But the Windows kernel module will be still unsigned right?

**Comment by zanderz** [comment]
@vagaerg We omitted the extra microsoft attestation signing step, but the version we've shipped in the last couple of weeks is properly signed. Sorry for the confusion.
